### PR TITLE
ledger consistency check

### PR DESCRIPTION
**Pull request by laizy**
ledger consistency check between nodes.

### CODE REVIEW

File: consensus/vbft/service.go [hunk 1]
```diff
@@ -124,6 +125,7 @@ type Server struct {
 	bftActionC chan *BftAction
 	msgSendC   chan *SendMsgEvent
 	sub        *events.ActorSubscriber
+	merkleRoot common.Uint256
```
Comment (Honglei-Cong) [inline consensus/vbft/service.go]: not used?

File: consensus/vbft/chain_store.go [hunk 2]
```diff
@@ -24,19 +24,26 @@ import (
...
+	"github.com/ontio/ontology/core/store"
+	"github.com/ontio/ontology/core/store/overlaydb"
+	"github.com/ontio/ontology/events/message"
 )
 
 type ChainStore struct {
 	db              *ledger.Ledger
 	chainedBlockNum uint32
 	pendingBlocks   map[uint32]*Block
+	execResult      *store.ExecuteResult
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: bind ExecuteResult with pendingBlocks.

type PendingBlock { *Block, *store.ExecuteResult }
change pendingBlock to map[uint32]PendingBlock

In the following Get/SetExecuteMerkleRoot(), change to Get/SetExecuteMerkleRoot(blockNum), 
and Get/Set ExecWriteSet

**Comment by laizy** [comment]
rebased to handle conflict

File: consensus/vbft/block_pool.go [hunk 3]
```diff
@@ -656,7 +656,7 @@ func (pool *BlockPool) setBlockSealed(block *Block, forEmpty bool) error {
 	}
 
 	// add block to chain store
-	if err := pool.chainStore.AddBlock(c.SealedBlock); err != nil {
+	if err := pool.chainStore.AddBlock(c.SealedBlock, pool.server); err != nil {
```
Comment (Honglei-Cong) [inline consensus/vbft/block_pool.go]: save *Server or Server.pid in chainStore when initializing chainStore?

File: consensus/vbft/types.go [hunk 4]
```diff
@@ -22,18 +22,20 @@ import (
...
 	vconfig "github.com/ontio/ontology/consensus/vbft/config"
 	"github.com/ontio/ontology/core/types"
-	"io"
 )
 
 type Block struct {
 	Block      *types.Block
 	EmptyBlock *types.Block
 	Info       *vconfig.VbftBlockInfo
+	MerkleRoot common.Uint256
```
Comment (Honglei-Cong) [inline consensus/vbft/types.go]: change to prevBlockMerkleRoot?

File: consensus/vbft/types.go [hunk 5]
```diff
@@ -56,6 +58,10 @@ func (blk *Block) getNewChainConfig() *vconfig.ChainConfig {
 	return blk.Info.NewChainConfig
 }
 
+func (blk *Block) getexecResMerkleRoot() common.Uint256 {
```
Comment (Honglei-Cong) [inline consensus/vbft/types.go]: getPrevBlockMerkleRoot()

File: core/store/ledgerstore/state_store.go [hunk 6]
```diff
@@ -43,24 +46,27 @@ var (
...
-	dbDir           string                    //Store file path
-	store           scom.PersistStore         //Store handler
-	merklePath      string                    //Merkle tree store path
-	merkleTree      *merkle.CompactMerkleTree //Merkle tree of block root
-	merkleHashStore merkle.HashStore
+	dbDir              string                    //Store file path
+	store              scom.PersistStore         //Store handler
+	merklePath         string                    //Merkle tree store path
+	merkleTree         *merkle.CompactMerkleTree //Merkle tree of block root
+	stateMerkleTree    *merkle.CompactMerkleTree //Merkle tree of state root
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/state_store.go]: rename to deltaMerkleTree?

File: consensus/vbft/service.go (hunk 1 above)
Comment (xiemylogos) [reply consensus/vbft/service.go]: has delete

File: consensus/vbft/block_pool.go (hunk 3 above)
Comment (xiemylogos) [reply consensus/vbft/block_pool.go]: now save *Server in chainstore in next commit

File: consensus/vbft/types.go (hunk 4 above)
Comment (xiemylogos) [reply consensus/vbft/types.go]: ok

File: consensus/vbft/chain_store.go (hunk 2 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: has fixed

**Comment by Honglei-Cong** [comment]
suggest making some stress tests

File: core/store/ledgerstore/ledger_store.go [hunk 7]
```diff
@@ -99,7 +105,7 @@ func NewLedgerStore(dataDir string) (*LedgerStoreImp, error) {
 	ledgerStore.blockStore = blockStore
 
 	stateStore, err := NewStateStore(fmt.Sprintf("%s%s%s", dataDir, string(os.PathSeparator), DBDirState),
-		fmt.Sprintf("%s%s%s", dataDir, string(os.PathSeparator), MerkleTreeStorePath))
+		fmt.Sprintf("%s%s%s", dataDir, string(os.PathSeparator), MerkleTreeStorePath), STATE_HASH_HEIGHT)
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go]: %s %s %s %d, 4 params to print

File: core/store/ledgerstore/ledger_store.go [hunk 8]
```diff
@@ -614,35 +738,32 @@ func (this *LedgerStoreImp) saveBlockToEventStore(block *types.Block) error {
...
 
-func (this *LedgerStoreImp) isSavingBlock() bool {
-	this.lock.Lock()
-	defer this.lock.Unlock()
-
-	if !this.savingBlock {
-		this.savingBlock = true
+func (this *LedgerStoreImp) tryGetSavingBlockLock() (hasLocked bool) {
+	select {
+	case this.savingBlockSemaphore <- true:
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go:755]: why not Mutex.TryLock()?

File: core/store/ledgerstore/ledger_store.go [hunk 9]
```diff
@@ -549,46 +613,106 @@ func (this *LedgerStoreImp) saveBlockToBlockStore(block *types.Block) error {
...
+		}
+
+		result.Notify = append(result.Notify, notify)
+	}
+
+	result.Hash = overlay.ChangeHash()
+	result.WriteSet = overlay.GetWriteSet()
+	if block.Header.Height < STATE_HASH_HEIGHT {
+		result.MerkleRoot = common.UINT256_EMPTY
+	} else if block.Header.Height == STATE_HASH_HEIGHT {
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go:650]: move the following {} to one separate function?

File: core/store/ledgerstore/ledger_store.go (hunk 7 above)
Comment (laizy) [reply core/store/ledgerstore/ledger_store.go]: `STATE_HASH_HEIGHT` is another param.

File: core/store/ledgerstore/ledger_store.go (hunk 9 above)
Comment (laizy) [reply core/store/ledgerstore/ledger_store.go:650]: done

**Comment by laizy** [comment]
rebased to fix conflict

File: consensus/vbft/chain_store.go [hunk 10]
```diff
@@ -24,22 +24,46 @@ import (
...
 	db              *ledger.Ledger
 	chainedBlockNum uint32
-	pendingBlocks   map[uint32]*Block
+	pendingBlocks   map[uint32]*PendingBlock
+	server          *Server
+	needSubmitBlock bool
 }
 
-func OpenBlockStore(db *ledger.Ledger) (*ChainStore, error) {
+func OpenBlockStore(db *ledger.Ledger, server *Server) (*ChainStore, error) {
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: suggest just adding serverPid to chainstore.

serverPid *actor.PID

File: consensus/vbft/chain_store.go [hunk 11]
```diff
@@ -48,13 +72,50 @@ func (self *ChainStore) GetChainedBlockNum() uint32 {
...
+	} else {
+		log.Errorf("SetExecMerkleRoot failed blkNum:%d,merkleRoot:%s", blkNum, merkleRoot.ToHexString())
+	}
+}
+
+func (self *ChainStore) GetExecMerkleRoot(blkNum uint32) common.Uint256 {
+	if blk, present := self.pendingBlocks[blkNum]; blk != nil && present {
+		return blk.execResult.MerkleRoot
+	}
+	merkleRoot, err := self.server.ledger.GetStateMerkleRoot(blkNum)
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: self.db == self.server.ledger

you can get MerkleRoot with self.db.GetStateMerkleRoot

File: consensus/vbft/chain_store.go [hunk 12]
```diff
@@ -24,22 +24,46 @@ import (
...
 		db:              db,
 		chainedBlockNum: db.GetCurrentBlockHeight(),
-		pendingBlocks:   make(map[uint32]*Block),
+		pendingBlocks:   make(map[uint32]*PendingBlock),
+		server:          server,
+		needSubmitBlock: false,
 	}, nil
 }
 
+func (self *ChainStore) Load() error {
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: move these to OpenChainStore() ?

File: consensus/vbft/service.go [hunk 13]
```diff
@@ -258,6 +263,11 @@ func (self *Server) LoadChainConfig(chainStore *ChainStore) error {
 	if err != nil {
 		return err
 	}
+	pendingBlock := &PendingBlock{block: block, execResult: &store.ExecuteResult{}}
```
Comment (Honglei-Cong) [inline consensus/vbft/service.go]: move these to OpenBlockStore()?

File: consensus/vbft/chain_store.go [hunk 14]
```diff
@@ -48,13 +72,50 @@ func (self *ChainStore) GetChainedBlockNum() uint32 {
 	return self.chainedBlockNum
 }
 
+func (self *ChainStore) SetExecMerkleRoot(blkNum uint32, merkleRoot common.Uint256) {
+	if blk, present := self.pendingBlocks[blkNum]; blk != nil && present {
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: suggest to change to setExecMerkleRoot()

File: consensus/vbft/chain_store.go [hunk 15]
```diff
@@ -48,13 +72,50 @@ func (self *ChainStore) GetChainedBlockNum() uint32 {
...
+		log.Errorf("GetStateMerkleRoot blockNum:%d, error :%s", blkNum, err)
+		return common.Uint256{}
+	} else {
+		return merkleRoot
+	}
+
+}
+
+func (self *ChainStore) SetExecWriteSet(blkNum uint32, memdb *overlaydb.MemDB) {
+	if blk, present := self.pendingBlocks[blkNum]; blk != nil && present {
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: suggest to change to setExecWriteSet()

Comment (Honglei-Cong) [reply consensus/vbft/chain_store.go]: If it is only used by OpenBlockStore(), suggest to remove the function, because this function update the PendingBlock **partially**.

File: consensus/vbft/chain_store.go [hunk 16]
```diff
@@ -78,58 +139,59 @@ func (self *ChainStore) AddBlock(block *Block) error {
...
+		}
+		self.pendingBlocks[blkNum] = &PendingBlock{block: block, execResult: &execResult}
+		self.needSubmitBlock = true
+		self.server.pid.Tell(
+			&message.BlockConsensusComplete{
+				Block: block.Block,
+			})
+		self.chainedBlockNum = blkNum
+		blkNum++
+		break
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: if always break, why need for{...}?
TO-BE-FIX

File: consensus/vbft/chain_store.go [hunk 17]
```diff
@@ -78,58 +139,59 @@ func (self *ChainStore) AddBlock(block *Block) error {
...
-func (self *ChainStore) SetBlock(block *Block, blockHash common.Uint256) error {
-
-	err := self.db.AddBlock(block.Block)
-	self.chainedBlockNum = self.db.GetCurrentBlockHeight()
-	if err != nil {
-		return fmt.Errorf("ledger failed to add block: %s", err)
-	}
-
-	return nil
+func (self *ChainStore) SetBlock(blkNum uint32, blk *PendingBlock) {
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store.go]: It seems SetBlock() only used when opening chain store.

File: consensus/vbft/msg_builder.go [hunk 18]
```diff
@@ -255,9 +255,10 @@ func (self *Server) constructProposalMsg(blkNum uint32, sysTxs, userTxs []*types
 
 	msg := &blockProposalMsg{
 		Block: &Block{
-			Block:      blk,
-			EmptyBlock: emptyBlk,
-			Info:       vbftBlkInfo,
+			Block:               blk,
+			EmptyBlock:          emptyBlk,
+			Info:                vbftBlkInfo,
+			PrevBlockMerkleRoot: self.chainStore.GetExecMerkleRoot(blkNum - 1),
```
Comment (Honglei-Cong) [inline consensus/vbft/msg_builder.go]: check if self.chainStore.GetExecMerkleRoot(blkNum - 1) failed?

File: consensus/vbft/msg_types.go [hunk 19]
```diff
@@ -357,7 +357,7 @@ func (msg *BlockFetchRespMsg) Deserialize(data []byte) error {
 	}
 	blk := &Block{}
 	if err := blk.Deserialize(buffer.Bytes()); err != nil {
-		return fmt.Errorf("unmarshal block type: %s", err)
+		return fmt.Errorf("Deserialize block type: %s", err)
```
Comment (Honglei-Cong) [inline consensus/vbft/msg_types.go]: seems unnecessary change

File: consensus/vbft/node_utils.go [hunk 20]
```diff
@@ -209,10 +209,13 @@ func (self *Server) buildParticipantConfig(blkNum uint32, block *Block, chainCfg
 	}
 
 	s := 0
-	cfg.Proposers = calcParticipantPeers(cfg, chainCfg, s, s+vconfig.MAX_PROPOSER_COUNT)
-	if uint32(len(cfg.Proposers)) < chainCfg.C {
+
+	Proposers := calcParticipantPeers(cfg, chainCfg, s, s+vconfig.MAX_PROPOSER_COUNT)
+	//cfg.Proposers = calcParticipantPeers(cfg, chainCfg, s, s+vconfig.MAX_PROPOSER_COUNT)
```
Comment (Honglei-Cong) [inline consensus/vbft/node_utils.go]: remove commented code?

File: consensus/vbft/types.go [hunk 21]
```diff
@@ -122,15 +130,22 @@ func (blk *Block) Deserialize(data []byte) error {
...
 	blk.Block = block
 	blk.EmptyBlock = emptyBlock
 	blk.Info = info
-
+	blk.PrevBlockMerkleRoot = merkleRoot
 	return nil
 }
 
-func initVbftBlock(block *types.Block) (*Block, error) {
+func initVbftBlock(block *types.Block, prevmerkleRoot common.Uint256) (*Block, error) {
```
Comment (Honglei-Cong) [inline consensus/vbft/types.go]: change param name to prevMerkleRoot ?

File: consensus/vbft/chain_store.go (hunk 10 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: has fixed

File: consensus/vbft/chain_store.go (hunk 11 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: ok

File: consensus/vbft/chain_store.go (hunk 12 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: ok

File: consensus/vbft/service.go (hunk 13 above)
Comment (xiemylogos) [reply consensus/vbft/service.go]: ok

File: consensus/vbft/chain_store.go (hunk 14 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: has modify

File: consensus/vbft/chain_store.go (hunk 15 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: has modify

File: consensus/vbft/chain_store.go (hunk 17 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: has delete

File: consensus/vbft/msg_types.go (hunk 19 above)
Comment (xiemylogos) [reply consensus/vbft/msg_types.go]: ok

File: consensus/vbft/node_utils.go (hunk 20 above)
Comment (xiemylogos) [reply consensus/vbft/node_utils.go]: ok

File: consensus/vbft/types.go (hunk 21 above)
Comment (xiemylogos) [reply consensus/vbft/types.go]: ok

File: consensus/vbft/msg_builder.go (hunk 18 above)
Comment (xiemylogos) [reply consensus/vbft/msg_builder.go]: has modify

File: consensus/vbft/chain_store.go (hunk 16 above)
Comment (xiemylogos) [reply consensus/vbft/chain_store.go]: has delete

**Review by Honglei-Cong — DISMISSED** [review]

File: consensus/vbft/chain_store_test.go [hunk 22]
```diff
@@ -54,7 +54,7 @@ func newChainStore() *ChainStore {
 func TestGetChainedBlockNum(t *testing.T) {
 	chainstore := newChainStore()
 	if chainstore == nil {
-		t.Error("newChainStrore error")
+		fmt.Println("newChainStore failed")
```
Comment (Honglei-Cong) [inline consensus/vbft/chain_store_test.go]: ???
The fix should be in newChainStore()?

Comment (xiemylogos) [reply consensus/vbft/chain_store_test.go]: ok

**Review by Honglei-Cong — DISMISSED** [review]

**Review by tanZiWen — DISMISSED** [review]

**Comment by laizy** [comment]
force push for rebase, change height to 1900000, other code is unchanged.
